### PR TITLE
Introduce HTTP input and output byte counters

### PR DESCRIPTION
**Pull request by wavesoft**
Summary:
This commit introduces a new filter that replaces the input and output streams
with a proxy stream that counts the bytes exchanged.

Related JIRA: MARATHON-8158

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    * order to count the number of bytes that went through the stream.
+    *
+    * @param upstream The stream to proxy the output to
+    */
+  private class ProxyOutputStream(val upstream: ServletOutputStream) extends ServletOutputStream {
+    var bytes: Int = 0
+
+    override def isReady: Boolean = upstream.isReady
+    override def setWriteListener(writeListener: WriteListener): Unit = upstream.setWriteListener(writeListener)
+    override def write(b: Int): Unit = {
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Are you sure that bytes are read/written one by one? What about overriding implementations of other read/write methods like `read(byte[] b)`, `read(byte[] b, int off, int len)`, `readLine(byte[] b, int off, int len)` or so on?

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  private[this] val inputBytesMetric = Metrics.counter(ServiceMetric, getClass, "inputBytes")
+  private[this] val outputBytesMetric = Metrics.counter(ServiceMetric, getClass, "outputBytes")
+
+  /**
+    * Wraps a `ServletOutputStream` and overrides the `write` method in
+    * order to count the number of bytes that went through the stream.
+    *
+    * @param upstream The stream to proxy the output to
+    */
+  private class ProxyOutputStream(val upstream: ServletOutputStream) extends ServletOutputStream {
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: I would maybe mention the word metrics or Measured or something, proxy is very generic, this is basically a Decorator to collect metrics...

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala (hunk 1 above)
Comment (wavesoft) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Well, the `read(): Int` and `write(b: Int)` are the only abstract methods of the base stream classes that **must** be implemented. Therefore I assumed that the default implementation of the other signatures is implemented by calling out to these base functions. And I confirmed this by checking the counters while polling various HTTP endpoints.

However, now that I think of it, the stream implementations _might_ provide a more efficient implementation for operations on byte ranges. So it actually makes sense to implement the other signatures of `read` and `write`. (But `readLine` does not really benefit from this.)

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala (hunk 2 above)
Comment (wavesoft) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: True, the term `Proxy` can be confusing 👍

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  private[this] val inputBytesMetric = Metrics.counter(ServiceMetric, getClass, "inputBytes")
+  private[this] val outputBytesMetric = Metrics.counter(ServiceMetric, getClass, "outputBytes")
+
+  /**
+    * Wraps a `ServletOutputStream` and overrides the `write` method in
+    * order to count the number of bytes that went through the stream.
+    *
+    * @param upstream The stream to proxy the output to
+    */
+  private class OutputStreamCounter(val upstream: ServletOutputStream) extends ServletOutputStream {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Well, I hate to break this to you but at least Apache Commons IO has this features: https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/input/CountingInputStream.html.

Comment (wavesoft) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: But the object I have to wrap is not an `OuptuStream`, it's a `ServletOutputStream` :trollface:

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Hm, there must be an easier way. `ServletOutputStream` is actually not doing much.

Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Ah man. I was so hopeful ^^.

Comment (timcharper) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: @jeschkies these other methods are necessary for async IO. See https://github.com/mesosphere/marathon/pull/6193/files#diff-52da5dcde3d226b130f1eff2c1c29c2b

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala (hunk 1 above)
Comment (timcharper) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: My understanding matches your suspicions, @wavesoft : all other methods call out to write(int).

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    override def read(b: Array[Byte], off: Int, len: Int): Int = {
+      bytes += len
+      upstream.read(b, off, len)
+    }
+  }
+
+  private class RequestCounterWrapper(val r: HttpServletRequest) extends HttpServletRequestWrapper(r) {
+    private lazy val stream = new InputStreamCounter(super.getInputStream)
+    def bytes: Int = stream.bytes
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Different name, perhaps? `bytesRead`, `bytesWritten`.

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import javax.servlet._
+import javax.servlet.http.{HttpServletRequest, HttpServletRequestWrapper, HttpServletResponse, HttpServletResponseWrapper}
+import mesosphere.marathon.metrics.{Metrics, ServiceMetric}
+
+/**
+  * This filter replaces the default I/O streams with proxies that count
+  * the number of bytes that are sent to and from the client.
+  */
+class HTTPMetricsFilter extends Filter {
+  private[this] val inputBytesMetric = Metrics.counter(ServiceMetric, getClass, "inputBytes")
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: `bytesRead`, `bytesWritten`?

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala (hunk 4 above)
Comment (wavesoft) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Not a bad idea 👍

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala (hunk 5 above)
Comment (wavesoft) [reply src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Not a bad idea 👍

File: src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    override def read(b: Array[Byte]): Int = {
+      val bytesRead = upstream.read(b)
+      if (bytesRead != -1) {
+        bytes += bytesRead
+      }
+      bytesRead
+    }
+
+    override def read(b: Array[Byte], off: Int, len: Int): Int = {
+      val bytesRead = upstream.read(b, off, len)
```
Comment (wavesoft) [inline src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala]: Ha! Nice catch 😄

**Pull request merged**